posts/2018-12-21-Procedural-Macros-in-Rust-2018.md

```diff
@@ -0,0 +1,358 @@
+The [`syn::parse` module][spm] has more information about this capability.
+
+### Producing a `TokenStream`
+
+Not only do we take a `TokenStream` as input with a procedural macro, but we
+also need to produce a `TokenStream` as output. This output is typically
+required to be valid Rust syntax, but like the input it's just list of tokens
+that we need to build somehow.
+
+Technically the only way to create a `TokenStream` is via its `FromIterator`
+implementation, which means we'd have to create each token one-by-one and
+collect it into a `TokenStream`. This is quite tedious, though, so let's take a
+look at [`syn`][syn]'s sibling crate: [`quote`].
+
+The [`quote`] crate is a quasi-quoting implementation for Rust which primarily
+provides a convenient macro for us to use:
+
+```rust
+use quote::quote;
+
+#[proc_macro_attribute]
+pub fn hello(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let input = syn::parse_macro_input!(input as syn::ItemFn);
+    let name = &input.ident;
+
+    // Our input function is always equivalent to returning 42, right?
+    let result = quote! {
+        fn #name() -> u32 { 42 }
+    };
+    result.into()
+}
+```
+
+The [`quote!` macro] allows you to write mostly-Rust syntax and interpolate
+variables quickly from the environment with `#foo`. This removes much of the
+tedium of creating a `TokenStream` token-by-token and allows quickly cobbling
+together various pieces of syntax into one return value.
+
+### Tokens and `Span`
+
+Perhaps the greatest feature of procedural macros in Rust 2018 is the ability to
+customize and use [`Span`] information on each token. A [`Span`] can be thought
+of as a pointer back into an original source file, typically saying something
+like "the `Ident` token` foo` came from file `bar.rs`, line 4, column 5, and was
+3 bytes long". This information is primarily used by the compiler's diagnostics
+with warnings and error messages.
+
+In Rust 2018 each [`TokenTree`] has a [`Span`] associated with it. This means that
+if you preserve the [`Span`] of all input tokens into the output then even
+though you're producing brand new syntax the compiler's error messages are still
+accurate!
+
+For example, a small macro like:
+
+```rust
+#[proc_macro]
+pub fn gen_fn(expr: TokenStream) -> TokenStream {
+    let result = quote! {
+        fn the_function() -> u32 { #expr }
+    };
+    result.into()
+}
+```
+
+when invoked as:
+
+```rust
+my_macro::gen_fn!("foo");
+```
+
+is invalid because we're returning a string from a function that should return a
+`u32`, and the compiler will helpfully diagnose the problem as:
+
+```
+error[E0308]: mismatched types
+ --> src/main.rs:1:19
+  |
+1 | my_macro::gen_fn!("foo");
+  | ------------------^^^^^--
+  | |                 |
+  | |                 expected u32, found reference
+  | expected `u32` because of return type
+  |
+  = note: expected type `u32`
+             found type `&'static str`
```

Comment:
This feels like not the *best* example perhaps -- it's somewhat an opaque error message. Perhaps there's something a little more nice-looking?

Reply:
Sure thing, I've pushed up a tweak